src/storage/rdb.cc

```diff
@@ -449,3 +510,188 @@ Status RDB::Restore(const std::string &key, uint64_t ttl_ms) {
+    } else if (type == RDBOpcodeFunction || type == RDBOpcodeFunction2) {
+      LOG(WARNING) << "RDB function not supported";
+      return {Status::NotOK, "RDB function not supported"};
+    } else {
+      if (!isObjectType(type)) {
+        LOG(WARNING) << "Invalid or Not supported object type: " << type;
+        return {Status::NotOK, "Invalid or Not supported object type"};
+      }
+    }
+
+    auto key = GET_OR_RETWITHLOG(LoadStringObject());
+    auto value = GET_OR_RETWITHLOG(loadRdbObject(type, key));
+
+    if (db_index != db_id) {  // skip db not match
+      continue;
+    }
+
+    if (isEmptyRedisObject(value)) {  // compatible with empty value
+      /* Since we used to have bug that could lead to empty keys
+       * (See #8453), we rather not fail when empty key is encountered
+       * in an RDB file, instead we will silently discard it and
+       * continue loading. */
+      if (empty_keys_skipped++ < 10) {
+        LOG(WARNING) << "skipping empty key: " << key;
+      }
+      continue;
+    } else if (expire_time != 0 &&
+               expire_time < now) {  // in redis this used to feed this deletion to any connected replicas
+      expire_keys++;
+      continue;
+    }
+
+    if (is_nx) {  // only load not exist key
+      auto s = exist(key);
+      if (!s.IsNotFound()) {
+        skip_exist_keys++;  // skip it even it's not okay
+        if (!s.ok()) {
+          LOG(ERROR) << "check key " << key << " exist failed: " << s.ToString();
+        }
+        continue;
+      }
+    }
+
+    auto ret = saveRdbObject(type, key, value, expire_time);
+    if (!ret.IsOK()) {
+      LOG(WARNING) << "save rdb object key " << key << " failed: " << ret.Msg();
+    } else {
+      load_keys++;
+    }
+  }
+
+  // Verify the checksum if RDB version is >= 5
+  if (rdb_ver >= 5) {
+    uint64_t chk_sum = 0;
+    auto expected = GET_OR_RETWITHLOG(stream_->GetCheckSum());
+    GET_OR_RETWITHLOG(stream_->Read(reinterpret_cast<char *>(&chk_sum), 8));
+    if (chk_sum == 0) {
+      LOG(WARNING) << "RDB file was saved with checksum disabled: no check performed.";
+    } else if (chk_sum != expected) {
+      LOG(WARNING) << "Wrong RDB checksum expected: " << chk_sum << " got: " << expected;
+      return {Status::NotOK, "Wrong RDB checksum"};
+    }
+  }
+
+  std::string skip_info = (is_nx ? ", exist keys skipped: " + std::to_string(skip_exist_keys) : "");
+
+  if (empty_keys_skipped > 0) {
+    LOG(INFO) << "Done loading RDB,  keys loaded: " << load_keys << ", keys expired:" << expire_keys
+              << ", empty keys skipped: " << empty_keys_skipped << skip_info;
+  } else {
+    LOG(INFO) << "Done loading RDB,  keys loaded: " << load_keys << ", keys expired:" << expire_keys << skip_info;
+  }
+
+  return Status::OK();
+}
+
+rocksdb::Status RDB::exist(const std::string &key) {
+  int cnt = 0;
+  std::vector<rocksdb::Slice> keys;
+  keys.emplace_back(key);
+  redis::Database redis(storage_, ns_);
+  auto s = redis.Exists(keys, &cnt);
```

Comment:
I was thinking of introducing the function `Database::KeyExist` to make the check, in this case, simpler without creating `std::vector` and renaming `Database::Exists` to `Database::MultipleKeysExist`. However, I'm not sure about it. @git-hulk @PragmaTwice What do you think?